File: 2/Solution.scala
def fibonacci(n: Int): Stream[Int] = {
  def fibs(a: Int = 0, b: Int = 1): Stream[Int] = Stream.cons(a, fibs(b, a+b))

  fibs() takeWhile(_ <= n)
}
println(fibonacci(3000) take 10 toList)
println(fibonacci(4000000).filter(_ % 2 == 0).sum)

File: 2/solution.rs
struct Fibonacci {
    curr: u32,
    next: u32,
}

impl Iterator for Fibonacci {
    type Item = u32;
    fn next(&mut self) -> Option<u32> {
        let new_next = self.curr + self.next;

        self.curr = self.next;
        self.next = new_next;

        Some(self.curr)
    }
}

fn fibonacci() -> Fibonacci {
    Fibonacci {curr: 1, next: 1}
}

fn main() {
    let result: u32 = fibonacci().take_while(|&x| x <= 4000000).filter(|x| x % 2 == 0).fold(0, |sum, x| sum + x);
    println!("{}", result.to_string())
}
